internal/cache/db.go

```diff
@@ -293,66 +281,64 @@ func openAndInitDB(path, dirtyFlagPath string) (*bbolt.DB, error) {
 
 // cleanExpiredUsers removes from the cache any user that exceeded the maximum amount of days without
 // online authentication.
-func (c *Cache) cleanExpiredUsers(expirationDate time.Time) (err error) {
+func (c *Cache) cleanExpiredUsers(expirationDate time.Time, procDir string) (err error) {
 	defer decorate.OnError(&err, "could not clean up database")
 
-	txError := c.db.Update(func(tx *bbolt.Tx) (txError error) {
+	return c.db.Update(func(tx *bbolt.Tx) (err error) {
 		// Cleans up the database
-		activeUsers, txError := getActiveUsers()
-		if txError != nil {
-			err = txError
-			return nil
+		activeUsers, err := getActiveUsers(procDir)
+		if err != nil {
+			return err
 		}
 
-		buckets, txError := getAllBuckets(tx)
-		if txError != nil {
-			err = txError
-			return nil
+		buckets, err := getAllBuckets(tx)
+		if err != nil {
+			return err
 		}
 
 		var expiredUsers []userDB
-		cursor := buckets[userByIDBucketName].Cursor()
-		for k, v := cursor.First(); k != nil; k, v = cursor.Next() {
+		// The ForEach call does not return an error, so we can ignore it.
+		_ = buckets[userByIDBucketName].ForEach(func(k, v []byte) error {
 			var u userDB
-			if err2 := json.Unmarshal(v, &u); err2 != nil {
-				err = errors.Join(err, err2)
-				continue
+			if err := json.Unmarshal(v, &u); err != nil {
+				slog.Warn(fmt.Sprintf("Could not unmarshal user %q: %v", string(k), err))
+				return nil
 			}
 
-			if _, ok := activeUsers[u.Name]; !ok && u.LastLogin.Before(expirationDate) {
+			if _, active := activeUsers[u.Name]; !active && u.LastLogin.Before(expirationDate) {
 				expiredUsers = append(expiredUsers, u)
 			}
-		}
-
-		if len(expiredUsers) == 0 {
-			slog.Debug("Skipping clean up as there are no expired users in the cache")
 			return nil
-		}
+		})
 
 		for _, u := range expiredUsers {
-			if err2 := deleteUser(buckets, u.UID); err2 != nil {
-				err = errors.Join(err, err2)
+			slog.Debug(fmt.Sprintf("Deleting expired user %q", u.Name))
+			if err := deleteUser(buckets, u.UID); err != nil {
+				slog.Warn(fmt.Sprintf("Could not delete user %q: %v", u.Name, err))
 			}
 		}
 
 		return nil
 	})
-
-	return errors.Join(err, txError)
 }
 
-// getActiveUsers walks through the /proc/ dir and returns a map with the usernames of the owners of all active processes.
-func getActiveUsers() (activeUsers map[string]struct{}, err error) {
+// getActiveUsers walks through procDir and returns a map with the usernames of the owners of all active processes.
+func getActiveUsers(procDir string) (activeUsers map[string]struct{}, err error) {
 	defer decorate.OnError(&err, "could not get list of active users")
 
 	activeUsers = make(map[string]struct{})
 
-	dirEntries, err := os.ReadDir("/proc/")
-	if err != nil {
+	dirEntries, err := os.ReadDir(procDir)
+	if err != nil || len(dirEntries) == 0 {
```

Comment:
sounds a little bit forced, would there be another way of having cleaning failing without exposing the procdir?

Reply:
I don't think so. The only points of failure in the cleanup function are:
```Go
activeUsers, err := getActiveUsers(procDir)
if err != nil {
	return err
}

buckets, err := getAllBuckets(tx)
if err != nil {
	return err
}
```

We could fail on getAllBuckets, but this would imply a corrupted database and there's no way we can get this in the first cleanup call (when creating the cache object).

Reply:
If you look at your coverage, it seems you triggered this error:
```
                info, err := dirEntry.Info()
                if err != nil {
                        return nil, err
                }
```